File: code_samples/scala/design_patterns_observer.scala
/**
 * Design Patterns: Observer
 * AI/ML Training Sample
 */

class Observer {
  private var data: String = ""
  
  def process(input: String): Unit = {
    data = input
  }
  
  def getData: String = data
  
  def validate: Boolean = data.nonEmpty
}

object ObserverExample {
  def main(args: Array[String]): Unit = {
    val instance = new Observer()
    instance.process("example")
    println(s"Data: ${instance.getData}")
    println(s"Valid: ${instance.validate}")
  }
}

File: code_samples/lua/design_patterns_observer.lua
-- Design Patterns: Observer
-- AI/ML Training Sample

Observer = {}
Observer.__ index = Observer

function Observer:new()
    local obj = {
        data = ""
    }
    setmetatable(obj, self)
    return obj
end

function Observer:process(input)
    self.data = input
end

function Observer:getData()
    return self.data
end

function Observer:validate()
    return #self.data > 0
end

-- Example usage
local instance = Observer:new()
instance:process("example")
print("Data: " .. instance:getData())
print("Valid: " .. tostring(instance:validate()))
